roachtest: add operation to probe ranges

noahstho opened the pull request:
Since SRE uses crdb_internal.probe_ranges to test for prod cluster health, we would like to add this as a roach operation
to make the DRT cluster as realistic as possible, and test for potential issues with crdb_internal.probe_ranges, so we know ASAP if our alerting coverage drops.

**Background**
crdb_internal.probe_ranges is a virtual table that quickly probes the entire keyspace of the KV layer to return a table of schema (range_id | error | end_to_end_latency_ms). It has minimal dependencies, so it functions even when a cluster is quite broken. And since it probes the entire keyspace, it is useful when something has already gone wrong, in narrowing down an issue to specific ranges.

**What will this roach operation can catch?**
If this roach operation fails, there is either a bug in `crdb_internal.probe_ranges`, so SRE is short a critical tool, or there is a serious bug is present in KV layer, and KV team will need to know asap. Ideally SRE would be first to know if there is an issue, and can hand off to KV if necessary.

**Testing PR**
Tested that it works on roachtest cluster with
`roachtest run-operation noahthompsoncockroachlabscom-test probe-ranges`, and also was able to test that it successfully failed by forcing a range_error in DB, w/
```./bin/roachtest run-operation noahthompsoncockroachlabscom-test2 probe-ranges

Running operation probe-ranges on noahthompsoncockroachlabscom-test2.

2025/04/29 20:00:46 run_operation.go:145: [1] operation status: checking if operation probe-ranges/read dependencies are met
2025/04/29 20:00:47 run_operation.go:145: [1] operation status: running operation probe-ranges/read with run id 12821170976295052991
2025/04/29 20:00:47 probe_ranges.go:92: [1] operation status: executing crdb_internal.probe-ranges read statement against node 3
2025/04/29 20:00:47 probe_ranges.go:92: [1] operation status: found 1 errors while executing crdb_internal.probe-ranges read statement against node 3
2025/04/29 20:00:47 probe_ranges.go:92: [1] operation status: error on node 3 on range 4: test range error
2025/04/29 20:00:47 operation_impl.go:138: [1] operation failure #1: Found range errors when probing via crdb_internal.probe-ranges read statement against node 3
2025/04/29 20:00:47 run_operation.go:229: recovered from panic: o.Fatal() was called
```

**Future Work**
We would like to also enable KVProber cluster setting to test this from a different angle, this should be a very easy change.

Fixes: https://github.com/cockroachdb/cockroach/issues/102034
Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144781)
<!-- Reviewable:end -->

noahstho committed:
roachtest: add operation to query crdb_internal.probe_ranges

Since SRE uses crdb_internal.probe_ranges to test for prod
cluster health, we would like to add this as a roach operation
to make the DRT cluster as realistic as possible, and test for
potential issues with crdb_internal.probe_ranges.

Fixes: #102034
Release note: None
Epic: None

DarrylWong commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 The Cockroach Authors.
```
nit: 2025

DarrylWong commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,57 @@
+	conn := c.Conn(ctx, o.L(), nid, option.VirtualClusterName(roachtestflags.VirtualCluster))
+	defer conn.Close()
+
+	probeReadRangeStmt := `select range_id, error from crdb_internal.probe_ranges(INTERVAL '1s', 'read')`
```
Consider if it makes sense to:

1. Randomize whether reads/writes are probed first.
2. Run the probes concurrently. i.e. how does the prober work via the scheduled prober? is it possible for the probes to run concurrently?

noahstho replied:
I split into two operations for read/write, as I understand operations can be run concurrently, so this should provide more interesting interleavings.

DarrylWong commented on pkg/cmd/roachtest/registry/owners.go:
```diff
@@ -35,6 +35,7 @@ const (
 	OwnerTestEng            Owner = `test-eng`
 	OwnerDevInf             Owner = `dev-inf`
 	OwnerFieldEng           Owner = `field-engineering`
+	OwnerSRE                Owner = `sre`
```
Actually after trying it locally it looks like it just fails the operation with: `probe-ranges: unknown owner "sre"`

noahstho replied:
Hmm, do you think it would make sense to add an SRE team, since we do own this? Or since github issues aren't created for operations, do you think there might be another team willing to take default ownership of operations?

DarrylWong replied:
Not too sure tbh, I'll defer judgement to DRP who would know if the ownership even matters here.

noahstho commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,57 @@
+
+	probeWriteRangeStmt := `select range_id, error from crdb_internal.probe_ranges(INTERVAL '1s', 'write')`
+
+	_, readErr := conn.ExecContext(ctx, probeReadRangeStmt)
```
Hmm, I think it might take longer if there's more data on cluster? not totally sure though.

Logging makes sense, I can add that and consider loop as well.

DarrylWong replied:
Yeah it should, I realized in my testing I didn't have a workload running in the background so it might not have been the most accurate test. You can always reiterate on this after seeing it run on a DRT cluster.

noahstho commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,57 @@
+
+	probeWriteRangeStmt := `select range_id, error from crdb_internal.probe_ranges(INTERVAL '1s', 'write')`
+
+	_, readErr := conn.ExecContext(ctx, probeReadRangeStmt)
```
The justification here is that we just want to see whether the virtual table is correctly created — if there's errors reported in the table that would definitely indicate a serious bug, but then I would be writing a test for KV not for the prober itself. There's also an existing known bug where we sometimes see spurious errors in the crdb_internal.probe_ranges on cluster startup IIRC, so not sure if that would actually pass...

DarrylWong replied:
> I would be writing a test for KV not for the prober itself

Right but you're already querying the errors so it seems to me that it'd be no extra cost to at least check them? Especially if you think it would indicate a serious bug; SRE doesn't necessarily have to triage them.

Maybe if not fail the operation, at least log the errors? I'm not too sure how noisy the false positives are.

noahstho replied:
Yes that makes sense to me — logging seems like a nice medium.

DarrylWong commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,70 @@
+	nodes := c.All()
+	nid := nodes[rng.Intn(len(nodes))]
+
+	conn := c.Conn(ctx, o.L(), nid, option.VirtualClusterName(roachtestflags.VirtualCluster))
```
Should this be `option.VirtualClusterName(install.SystemInterfaceName))` instead? I know invalid descriptor checks have to be run on the system level, not sure if the same is true for `probe_ranges`.

noahstho replied:
hmmmm.... i am not sure... going to leave this open and look into it

or maybe it will be obvious to DRP team.

shailendra-patel replied:
On DRT cluster we don't run virtual clusters. So this flag value is empty while running the operations. Therefore eventually it is going to connect to system tenant. It is fine i guess the way it is.

DarrylWong commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,70 @@
+
+// runProbeRanges runs a read or write stmt against crdb_internal.probe_ranges.
+func runProbeRanges(
+	ctx context.Context, o operation.Operation, c cluster.Cluster, write_probe bool,
```
super nit: writeProbe instead of `write_probe`

DarrylWong reviewed (approved):
LGTM, although I'd get a stamp from someone in DRP or SRE before merging

noahstho commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+func registerProbeRanges(r registry.Registry) {
+	r.AddOperation(registry.OperationSpec{
+		Name:               "probe-ranges/read",
+		Owner:              registry.OwnerKV,
```
DRP team could you please advise? I would not like to page KV team for this operation, but there is no SRE team currently set up...

shailendra-patel replied:
This will not trigger a page to KV. The owner flag is purely informational, indicating which team should be contacted if this operation fails.

DarrylWong commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+func runProbeRanges(
+	ctx context.Context, o operation.Operation, c cluster.Cluster, writeProbe bool,
+) registry.OperationCleanup {
+	rng, _ := randutil.NewPseudoRand()
```
an aside: it doesn't seem ideal for operations to be managing their own RNG. e.g. if we want to reproduce a specific operation run, it's up to each individual operation to make sure to log the seed which we aren't doing here.

Instead, the top level operation runner should be calling `randutil.NewPseudoRand()` which uses that to generate+log a new seed for each individual operation.

Anyway, outside the scope of this PR, just mentioning it.

shailendra-patel commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+
+	o.Status(fmt.Sprintf("executing crdb_internal.probe-ranges %s statement against node %d", probeType, nid))
+
+	probeRangeStmt := fmt.Sprintf(`select count(error) as errors from crdb_internal.probe_ranges(INTERVAL '1s', '%s')`, probeType)
```
On my roachprod cluster when i run this query i get all the records. The value of error column is set to empty, therefore the query is giving all the records.
```
roachprod@localhost:26257/defaultdb> select count(error) as errors from crdb_internal.probe_ranges(INTERVAL '1s', 'read');                                                                               
-[ RECORD 1 ]
errors | 54
```

The correct query would be
```
select count(1) from crdb_internal.probe_ranges(INTERVAL '1s', 'read') where error != ''
```
The above query is referenced from [this code](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/generator_probe_ranges.go#L57)

shailendra-patel commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+
+	var errors int
+	if scan_err := range_errors.Scan(&errors); scan_err != nil {
+		o.Status(fmt.Sprintf("found %d errors while executing crdb_internal.probe-ranges %s statement against node %d", errors, probeType, nid))
```
If there is any error in reading value from the db row using scan we should fail the operation with msg like `failed to scan value from crdb_internal.probe_ranges`. Use `o.Fatalf` for failing the operation

shailendra-patel commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+	if scan_err := range_errors.Scan(&errors); scan_err != nil {
+		o.Status(fmt.Sprintf("found %d errors while executing crdb_internal.probe-ranges %s statement against node %d", errors, probeType, nid))
+	}
+
```
If there are entries with error column in virtual table `crdb_internal.probe_ranges` does it mean something wrong is happening in the cluster which needs attention? If yes, we should  check if `errors > 0`  and fail the operations. 

If it is common to have entries with error in `crdb_internal.probe_ranges`  it is better to have a threshold value for error and fail the operation if error exceeds the threshold.

noahstho replied:
I was originally intending this code change to only ensure that the virtual table correctly works, but you raise a good point. I have changed to do as you suggest — although if we do see range errors, it is likely many other things will be failing.

shailendra-patel commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,78 @@
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
```
It would be nice if you can add more information about this operation. Answering following question would help:
1. What data does `crdb_internal.probe_ranges` table contains.
2. What does it mean if cluster is observing failed read/write probes.
3. Corrective action needed in case this operation fails?

noahstho replied:
Hi, thanks for feedback, I will add some more information to file and make fixes you suggested. In meantime here are answers to your questions in a more detailed manner:

1. `crdb_internal.probe_ranges` is a virtual table  that quickly probes the _entire_ keyspace of the KV layer to return a table of schema (range_id | error | end_to_end_latency_ms). It has minimal dependencies, so it functions even when a cluster is quite broken. And since it probes the _entire_ keyspace, it is useful when something has already gone wrong, in narrowing down an issue to specific ranges.

There is a lot of conceptual overlap and similarity with the KVProber which cheaply samples and probes _random_ keys in the keyspace. However the two are implemented separately, and the KVProber is primarily used for alerting, rather than debugging. As a side note, I would also like to eventually enable the KVProber cluster setting on DRT clusters as well.

2. If the cluster is failing read/write probes, it should mean that certain ranges are broken, which would likely mean that a ton of other operations would also be failing. Range unavailability would likely produce failures on the SQL level in general. We have not extensively tested `crdb_internal.probe_ranges` on a long lived cluster, however, so it's possible we might uncover a bug though. It would be very interesting if we observed behavior with `crdb_internal.probe_ranges` failing but everything else working.

3. In the case this operation fails, there are two possibilities, either there is an underlying issue with range unavailability, or there is an issue with `crdb_internal.probe_ranges` itself, ex: [this motivating bug](https://github.com/cockroachdb/cockroach/issues/101549). If there's an issue with `crdb_internal.probe_ranges` then SRE or KV needs to fix it, since it's a critical tool for incident response. If there's an issue with range unavailability, then that means that we've also probably uncovered a serious bug, so SRE or KV need to do some investigative work to fix that as well.


original github issue for reference: https://github.com/cockroachdb/cockroach/issues/61695

noahstho commented:
@shailendra-patel Thank you for review, and apologies for not providing enough context in description. I have responded to questions and added some more background info to the PR, as well as making the suggested change to fail if we detect range errors.

PTAL and let me know if any general questions still remain on the overall approach.

shailendra-patel commented on pkg/cmd/roachtest/operations/probe_ranges.go:
```diff
@@ -0,0 +1,108 @@
+
+	var errors int
+	if !range_errors.Next() {
+		o.Fatalf(fmt.Sprintf("Failed to scan value from crdb_internal.probe-ranges %s statement against node %d", probeType, nid))
```
nit: `o.Fatalf` already handles formatting internally, so `fmt.Sprintf` is redundant. 
Same for other usage of `o.Fatalf`

noahstho commented:
bors r+
